### PR TITLE
Add subtreeID, depositCount to DepositQueued event

### PR DESCRIPTION
**Pull request by jacque006**
Resolves https://github.com/thehubbleproject/hubble-contracts/issues/637

### CODE REVIEW

File: contracts/DepositManager.sol [hunk 1]
```diff
@@ -83,9 +95,13 @@ contract DepositCore is SubtreeQueue {
             i >>= 1;
         }
         babyTreesLength = len;
+        // TODO Is there a better way we can mirror this as it is in enqueue function?
+        subtreeID = back + 1;
```
Comment (jacque006) [inline contracts/DepositManager.sol:99]: Is there a better way we can do this to reduce duplication of next subtreeID being calculated?

File: ts/client/services/events/depositQueued.ts [hunk 2]
```diff
@@ -13,7 +13,7 @@ export class DepositQueuedEventSyncer extends ContractEventSyncer {
     constructor(api: CoreAPI) {
         super(
             api.contracts.depositManager,
-            api.contracts.depositManager.filters.DepositQueued(null, null, null)
+            api.contracts.depositManager.filters.DepositQueued()
```
Comment (jacque006) [inline ts/client/services/events/depositQueued.ts:16]: Newer typechain/ethers versions do not require `null` filling. Will update this where appropriate in future work, and may do a tech-debt PR after public testnet to go through client/tests and update all of these.

**Comment by jacque006** [comment]
Gas changes

### master
Screenshot from 2021-07-29 11-39-53
Deposit 0: `139601`
Deposit 1: `109921`

### This PR
Screenshot from 2021-07-29 11-40-40
Deposit 0: `142333`
Deposit 1: `110665`
Deposit 2: `108145`

### cha-cha-cha-changes

Deposit 0 change: `+2732`
Deposit 0 % change: `1.957%`

Deposit 1 change: `+744`
Deposit 1 % change: `0.677%`

Deposit 2 N/A

File: contracts/DepositManager.sol (hunk 1 above)
Comment (ChihChengLiang) [reply contracts/DepositManager.sol:99]: Yes, let me reply in other comments

File: contracts/DepositManager.sol [hunk 3]
```diff
@@ -65,7 +74,10 @@ contract DepositCore is SubtreeQueue {
         paramMaxSubtreeSize = 1 << maxSubtreeDepth;
     }
 
-    function insertAndMerge(bytes32 depositLeaf) internal {
+    function insertAndMerge(bytes32 depositLeaf)
+        internal
+        returns (uint256 subtreeID, uint256 prevDepositCount)
+    {
         depositCount++;
         uint256 i = depositCount;
```
Comment (ChihChengLiang) [inline contracts/DepositManager.sol:82]: We make good use of the loaded depositCount here.
```suggestion
        depositID = depositCount + 1; // 1 SLOAD here
        uint256 i = depositID;
```

File: contracts/DepositManager.sol [hunk 4]
```diff
@@ -83,9 +95,13 @@ contract DepositCore is SubtreeQueue {
             i >>= 1;
         }
         babyTreesLength = len;
+        // TODO Is there a better way we can mirror this as it is in enqueue function?
+        subtreeID = back + 1;
+        prevDepositCount = depositCount;
```
Comment (ChihChengLiang) [inline contracts/DepositManager.sol:100]: Don't do them here yet.
```suggestion
```

File: contracts/DepositManager.sol [hunk 5]
```diff
@@ -65,7 +74,10 @@ contract DepositCore is SubtreeQueue {
         paramMaxSubtreeSize = 1 << maxSubtreeDepth;
     }
 
-    function insertAndMerge(bytes32 depositLeaf) internal {
+    function insertAndMerge(bytes32 depositLeaf)
+        internal
+        returns (uint256 subtreeID, uint256 prevDepositCount)
```
Comment (ChihChengLiang) [inline contracts/DepositManager.sol:79]: Rename to make sense the following feedbacks.
```suggestion
        returns (uint256 subtreeID, uint256 depositID)
```

File: contracts/DepositManager.sol [hunk 6]
```diff
@@ -83,9 +95,13 @@ contract DepositCore is SubtreeQueue {
...
+        subtreeID = back + 1;
+        prevDepositCount = depositCount;
+
         // Subtree is ready, send to SubtreeQueue
         if (depositCount == paramMaxSubtreeSize) {
-            enqueue(babyTrees[0]);
+            subtreeID = enqueue(babyTrees[0]);
             // reset
             babyTreesLength = 0;
             depositCount = 0;
```
Comment (ChihChengLiang) [inline contracts/DepositManager.sol:107]: Do early return here.
```suggestion
            depositCount = 0;
            return
```
Then after the "if" block, we do 
```
// The case of no reset
subtreeID = back + 1;
// bump the depositCount
depositCount = depositID;
```

File: contracts/DepositManager.sol [hunk 7]
```diff
@@ -83,9 +95,13 @@ contract DepositCore is SubtreeQueue {
             i >>= 1;
         }
         babyTreesLength = len;
+        // TODO Is there a better way we can mirror this as it is in enqueue function?
+        subtreeID = back + 1;
+        prevDepositCount = depositCount;
+
         // Subtree is ready, send to SubtreeQueue
         if (depositCount == paramMaxSubtreeSize) {
```
Comment (ChihChengLiang) [inline contracts/DepositManager.sol:103]:
```suggestion
        if (depositID == paramMaxSubtreeSize) {
```

**Review by ChihChengLiang — COMMENTED** [review]
Proposed a way to assign subtreeID gracefully.

**Comment by jacque006** [comment]
Thanks for the help @ChihChengLiang ! Suggestions implemented, gas stats updated, slight decrease :tada:

**Review by ChihChengLiang — COMMENTED** [review]
LGTM

**Comment by kautukkundan** [comment]
Suggestion - we can start using [eth-gas-reporter](https://github.com/cgewecke/eth-gas-reporter/). It also comes with [codechecks](https://github.com/cgewecke/eth-gas-reporter/blob/master/docs/codechecks.md) plug-in which can compare gas change with the target branch so we don't need to calculate manually!

**Comment by jacque006** [comment]
@kautukkundan I have a branch up that would add contract test coverage as well as hopefully implement the reporter you mentioned https://github.com/thehubbleproject/hubble-contracts/pull/552 . We would need to update a number of spots for it to work properly but I agree that it would be much easier than manually calculating and reporting in PRs :)

**Pull request merged**